src/ostrowski/rationals/bounded.lean

```diff
@@ -0,0 +1,364 @@
+import data.real.cau_seq
+
+import ostrowski.rationals.basic
+import abvs_equiv
+
+open list is_absolute_value
+
+lemma prime_norm_lt_one_of_bounded_padic (abv : ℚ → ℝ) [habv : is_absolute_value abv]
+      (hnontriv: abv ≠ trivial_abs)
+      (all_nat_le_one: (∀ z : ℕ, abv z ≤ 1)):
+      ∃ p: ℕ, prime p ∧ abv p < 1 :=
+begin
+  by_contra,
+  push_neg at h,
+
+  suffices h': ∀ n: ℕ, n ≠ 0 → abv n = 1,
+  {
+    apply hnontriv,
+    set abv_hom := hom_of_abv abv with abv_hom_def,
+    set triv_hom := hom_of_abv trivial_abs with triv_hom_def,
+    have abv_hom_f: abv = abv_hom := by refl,
+    have triv_hom_f: trivial_abs = triv_hom := by refl,
+    rw [abv_hom_f, triv_hom_f],
+    apply mul_mor_eq_of_eq_on_pnat,
+    {
+      rw [← abv_hom_f, ← triv_hom_f],
+      rw [abv_neg abv, abv_neg trivial_abs],
+      rw [abv_one abv, abv_one trivial_abs],
+    },
+    {
+      intro n,
+      rw [← abv_hom_f, ← triv_hom_f],
+      unfold trivial_abs,
+      by_cases h'': n = 0,
+      {
+        rw h'',
+        simp,
+        rw abv_zero abv,
+      },
+      {
+        rw h' n h'',
+        simp [h''],
+      }
+    },
+  },
+  
+  apply @induction_on_primes (λ n, n ≠ 0 → abv n = 1),
+  { tauto },
+  { intro h, exact abv_one abv, },
+  {
+    intros p a p_prime ha hprod,
+    push_cast,
+    rw abv_mul abv,
+    have abvp_eq_one: abv p = 1,
+    { by linarith [h p p_prime, all_nat_le_one p], },
+    rw abvp_eq_one,
+    have a_ne_zero: a ≠ 0,
+    { by_contra h, push_neg at h, apply hprod, simp [h], },
+    rw ha a_ne_zero,
+    simp,
+  },
+end
+
+lemma padic_norm_q (p: ℕ) (q: ℕ) (p_prime: prime p) (q_prime: prime q):
+  (p = q → padic_norm p q = 1/p) ∧ (p ≠ q → padic_norm p q = 1) :=
+begin
+  split,
+  {
+    intro h, -- In the case where `p = q`
+    have p: padic_val_rat p q = 1,
+    {
+      rw ← h,
+      unfold padic_val_rat,
+      simp [p_prime.1, prime.ne_one p_prime],
+    },
+    unfold padic_norm,
+    rw p,
+    simp [q_prime.1],
+  },
+  {
+    intro h, -- In the case where `p ≠ q`
+    have p: padic_val_rat p q = 0,
+    {
+      unfold padic_val_rat,
+      simp [q_prime.1, prime.ne_one p_prime],
+      suffices h: p^0 ∣ q ∧ ¬ p^1 ∣ q,
+      {
+        norm_cast,
+        rw roption.get_eq_iff_eq_some,
+        rw multiplicity.eq_some_iff.2 h,
+        refl,
+      },
+      split,
+      {
+        simp,
+      },
+      {
+        by_contra p_div_q,
+        apply h,
+        rw ← associated_iff_eq,
+        simp at p_div_q,
+        have q_div_p := dvd_symm_of_irreducible
+          (irreducible_of_prime p_prime)
+          (irreducible_of_prime q_prime)
+          p_div_q,
+        exact associated_of_dvd_dvd p_div_q q_div_p,
+      },
+    },
+    unfold padic_norm,
+    rw p,
+    simp [q_prime.1],
+  },
+end
+
+lemma abs_val_bounded_q (abv : ℚ → ℝ) [habv : is_absolute_value abv]
+  (all_nat_le_one: (∀ z : ℕ, abv z ≤ 1))
+  (p: ℕ) (p_prime: prime p) (abvp_lt_one: abv p < 1):
+  ∃ α: ℝ, 0 < α ∧ ∀ q: ℕ, prime q →
```

Comment:
split the conclusion into two lemmas, and generate aux lemmas for preliminaries before the actual use

Reply:
Actually I plan on greatly simplifying this proof, but it will require some valuation theory so I leave it as it is for now.